Added a blogpost on the experince of running nf-core pipelines in google colab (#3471)
nf-core/website: sites/main-site/src/content/blog/2025/nf-core-colab-guide.mdx

This means you will have access to all your favorite extensions and syntax highlighting in a familiar, seamless GUI!
The library makes use of the official [VS Code Remote Tunnels](https://code.visualstudio.com/docs/remote/tunnels) to securely and reliably connect Google Colab as well as Kaggle notebooks to a local or browser-based instance of VS Code.
You can read more about the library and even help contribute to new features on its [GitHub repository](https://github.com/EssenceSentry/vscode-colab).

# Downsides and Caveats of the VS Code Colab Approach

**Main limitations:**

- No root access (no Docker/Singularity)
- Session timeouts
- MPLBACKEND issues
- Conda is not native
- Limited GUI for complex workflows

While this approach is great, it does have it's downsides.
The biggest issue you will face is frequent disconnections or crashing of the connection tunnel.
I have seen that this reduces drastically—or almost never happens—if you make sure to set up the other aspects of your Colab environment before starting up the tunnel.

> The issue could be with Colab having a low affinity for handling requests from multiple frontends.

Author: Colab struggles to handle simultaneous queries from multiple instances of the same VM, the reason could be that it wasn't designed with this in mind

At the time of writing this blog post, the most annoying issue is that you have to set up the whole VS Code environment with all the extensions from scratch with each run.
The developer of the package did indicate that the ability to save profiles as config files is under development and will be added soon, so make sure to keep an eye on the repo for any such developments.
Although not a huge issue, I found the tunnel construction time of 3-5 minutes to be a bit too long to wait.
Other than these, the package works great and just about seamlessly gets the job done.

# Tips for a Smooth Experience

As an important step before running your pipeline, or any others in Colab, is to set the `MPLBACKEND` environment variable to `Agg`.

> This ensures that Matplotlib can render plots in a headless environment like Colab, where no display is available.

Author: I don't know, but it's a crash that happens every time I try to run a pipeline without doing this first. I guess even if we don't get a direct plot output, any kind of rendering and saving to memory of plots handled by Matplotlib is affected by the issue.

> This ensures that Matplotlib can render plots in a headless environment like Colab, where no display is available.

Reviewer: Huh... do you know which step of the pipelines it's happening on ? Do you the error still somewhere or can you re-generate it?

You can do this either by running the following in a code cell:

```python title="Set MPLBACKEND to Agg in a code cell"
%env MPLBACKEND=Agg
```

Or alternatively, by running the following command in the terminal:

```bash title="Set MPLBACKEND to Agg in the terminal"
export MPLBACKEND=Agg
```

Google colab 's storage is temporary and also limited in most cases.
As such, make it standard practice to mount and save your results in your personal Google Drive account if you have one.
Additionally, if you plan on writing and developing your pipelines exclusively in Google Colab, then make sure to work in git and commit, or alternatively test in Colab but save changes from your local PC and commit to prevent loss of your progress.
Finally, make sure to pick the VM instance that works best for your task and set up your Nextflow run configuration file accordingly to make the most use of the hardware at your disposal.

> If you have feedback, questions, or tips, please share them in the comments or via the nf-core Slack channel. Your input helps improve the community!

Author: I meant the workspace in general. Is there a channel that is well suited to handling this?

---
title: "Running nf-core Pipelines on Google Colab"
subtitle: "A guide to running and interacting with nf-core pipelines using Colab and VS Code"
pubDate: 2025-08-20T12:00:00+01:00
headerImage: "/assets/images/blog/nf-core-colab/blogpost-thumbnail.jpg"
headerImageAlt: "Epic Handshake between nf-core and colab"
embedHeaderImage: true
authors:
    - "KurayiChawatama"
label:
    - "community post"
---

import { Image } from "astro:assets";
import Profile from "@components/GitHubProfilePictureExtended.svelte";
import colabmeme from "@assets/images/blog/nf-core-colab/colab-meme.jpg";
import disconnectmeme from "@assets/images/blog/nf-core-colab/colab-disconnected-meme.png";
import sleepmeme from "@assets/images/blog/nf-core-colab/dont-fall-asleep-meme.png";

> ## Running nf-core Pipelines in Google Colab

Reviewer 2: ```suggestion
## Running nf-core pipelines in Google Colab
```

Running and developing nf-core pipelines can be computationally intensive, requiring resources not easily available to students, newcomers, or participants in hands-on training workshops.

> Google Colab provides a free, accessible way to leverage powerful cloud hardware for computational tasks, making it an attractive option for students, researchers, and anyone with limited local resources.

Reviewer 3: ```suggestion
Google Colab provides an affordable and sometimes even free, accessible way to leverage powerful cloud hardware for computational tasks, making it an attractive option for students, researchers, and anyone with limited local resources.
```

At least if I log in with my SciLifeLab account to Colab, I get the notification that any usage will be billed to my organization. If we stress the _free_ too much, people may complain to us if they are afterward billed unexpectedly.

To make it easier for people to access such resources, we have just published a new detailed tutorial for running and developing nf-core/Nextflow pipelines in Google Colab is available at [this link](https://nf-co.re/docs/tutorials/google_colab/nf-core-colab-guide)!

In this blog post we provide background into our own experiences of running and developing Nextflow and nf-core pipelines, describing its pros and cons, but that also motivated the creation of the tutorial.

> ### Why Run Pipelines in Google Colab

Reviewer 2: ```suggestion
### Why run pipelines in Google Colab
```

<Image
    src={colabmeme}
    class="d-block m-auto"
    width={600}
    density={[1.5, 2]}
    alt="Butch eats GPUs from the garbage while Tom is handfed in luxury"
/>

> Google Colab provides free credits for basic computational infrastructure, making it an interesting option for people with limited resources to run computational workflows.

Author: more accurate actually, don't know how this got there

> Furthermore, for a fee, you can access the latest hardware each year at a fraction of the cost of buying a new PC.

Reviewer 3: ```suggestion
```

Author: This is my rationale, actually, by my calculations, I will spend around 300 USD  in monthly colab payments for my 4 college years to do my GPU-heavy computational work. They always have the latest GPUs available for use, which beats buying myself a new GPU every year

Reviewer 3: I reckoned nobody will buy a new GPU every year and very few nf-core pipelines actually profit of GPUs at the moment, but since it is in your blog post, I am fine with keeping it as well!

As bioinformatics and data science tasks become more resource-intensive, Colab offers a potentially cost-effective solution for learning and developing large-scale pipelines like those built with Nextflow.
If you’re a student or developer working on a laptop, Colab can dramatically speed up your workflow.
For example, you can write and test your code locally, then run it in Colab to take advantage of faster execution and larger memory—saving time and reducing frustration from crashes on limited hardware.

Colab is also ideal for training workshops where participants may not have access to high-performance computing (HPC) clusters.
Instructors can use real-world, large datasets instead of toy examples, thus giving everyone hands-on experience with industry-scale workflows.

> Ultimately, using Colab can help democratize access to advanced pipeline development and best practices, enabling more people to contribute to open-source projects like nf-core.

Reviewer 3: More a question to the maintainers: Since nf-core's Code of Conduct specifically mentions socioeconomic status, it could be mentioned in this context?

Thankfully, the [vscode-colab](https://github.com/EssenceSentry/vscode-colab) Python library provides just the toolkit you need to take advantage of Colab's powerful hardware in the comfort of VS Code's rich software suite.
This means you will have access to all your favorite extensions and syntax highlighting in a familiar, seamless GUI!
The library makes use of the official [VS Code Remote Tunnels](https://code.visualstudio.com/docs/remote/tunnels) to securely and reliably connect Google Colab as well as Kaggle notebooks to a local or browser-based instance of VS Code.

You can read more about the library and even help contribute to new features on its [GitHub repository](https://github.com/EssenceSentry/vscode-colab).

### Limitations of the VS Code Colab approach

**Main limitations:**

- No root access (no Docker/Singularity)
- Session timeouts
- `MPLBACKEND` issues
- Conda is not native
- Limited GUI for complex workflows

While the `vscode-colab` approach is great, it does have it's downsides.

The biggest issue you will face is frequent disconnections or crashing of the connection tunnel.

> I have seen that if you make sure to set up the other aspects of your Colab environment before starting up the tunnel - disconnections rarely happens (or at least drastically reduces the number of).

Reviewer 3: Mind that you started the blog post with we, are switching to I in the middle section and getting back to we in the end. Whatever you decide for, I recommend keeping it consistent.

> In this blog post, we provide background into our own experiences of running and developing Nextflow and nf-core pipelines

> Although not a huge issue, I found the tunnel construction time of 3-5 minutes to be a bit too long to wait.

> We hope this blog post and the tutorial will help kick-start the community members interested in trying out Google Colab for their own nf-core work!

Apart from that, it jumps back and forth between personal experiences in past tense and generalized propositions in present tense. While this is not strictly wrong, I wonder if a clearer separation or rephrasing some propositions as personal experiences as well would help the reading flow?

Reviewer: Sorry, I/we thing was probably my fault with my suggestions!
